core/executor/src/wasm_executor.rs

```diff
@@ -84,7 +56,7 @@ impl<'e, E: Externalities<Blake2Hasher>> FunctionExecutor<'e, E> {
 	fn new(m: MemoryRef, t: Option<TableRef>, e: &'e mut E) -> Result<Self> {
 		Ok(FunctionExecutor {
 			sandbox_store: sandbox::Store::new(),
-			heap: Heap::new(&m),
+			heap: heap::Heap::new(m.used_size().0 as u32),
```

Comment:
It seems that `used_size` is treated as a maximum size of a heap. However, this is not accurate since `used_size` is the highest address that was touched in this memory instance so far. 

It was used as some sort of heuristic to determine a start pointer for a bump allocator, assuming that after the initialization (i.e. copying the data segments onto the memory and running a special initializer function) was done, all memory that is touched is free and is available for layouting.

Reply:
I see, how about using `current_size` (https://github.com/paritytech/wasmi/blob/master/src/memory.rs#L182) instead?

Reply:
Sorry, I haven't yet done a proper review, so here are some questions (maybe candidates for comments/rustdocs ; ) ):

1. What this `reserved` parameter is supposed to represent?
2. Does this allocator has an upper limit (akin we used to had with heap pages)?
3. If I allocate memory for the first time, what address the allocator will return?

Reply:
`used_size` should be treated as starting base heap address. `Heap` should start allocating from this address, not from `0`. `current_size` would be the maximum heap size.

Reply:
@arkpar The heap takes only one argument, which is the total size it can use. When allocating it returns an index offset from `1…reserved - buddyHeapBlocksize`.

@pepyakin 
1. The reserved parameter is the maximum heap size, i.e. the `reserved` space of memory which is available.
2. Yes, the upper limit is this `reserved` parameter. The implementation will return `0` if it's exhausted.
3. `1`, there is also a test for it: https://github.com/paritytech/substrate/blob/cmichi-better-allocator-for-wasm/core/executor/src/heap.rs#L271-L276.

Reply:
@arkpar Thanks for the `current_size` comment, will adapt the code.

Reply:
3. I'm afraid this is not correct. As @arkpar mentioned, `Heap` shouldn't start returning addresses neither from 0 or 1. Wasm binary can (and do) reserve some memory for it's own purposes (like, stack and data section) and the worst thing is that we even can't tell what memory is reserved. But we can guess and `used_size` does exactly that (see my first comment)!

Reply:
Thanks for the explanation! I updated the PR accordingly.